### PR TITLE
[Model Registry Tags] Python client implementation and tests

### PR DESCRIPTION
**Pull request by zhidongqu-db**
## What changes are proposed in this pull request?

Python client implementation and tests for model registry tags

## How is this patch tested?

Unit tests, integration tests, manual tests

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for
Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tracking/_model_registry/client.py [hunk 1]
```diff
@@ -28,17 +29,19 @@ def __init__(self, registry_uri):
 
     # Registered Model Methods
 
-    def create_registered_model(self, name):
+    def create_registered_model(self, name, tags=None):
         """
         Create a new registered model in backend store.
 
         :param name: Name of the new model. This is expected to be unique in the backend store.
+        :param tags: A list of :py:class:`mlflow.entities.model_registry.RegisteredModelTag`
```
Comment (dbczumar) [inline mlflow/tracking/_model_registry/client.py:37]: For run tags, client-level APIs accept a dictionary of tags and converts them to a list of `RunTag` objects before passing them to the backend store (see https://github.com/mlflow/mlflow/blob/master/mlflow/tracking/_tracking_service/client.py#L88). I think we should do the same thing here.

File: mlflow/tracking/_model_registry/client.py [hunk 2]
```diff
@@ -128,19 +131,42 @@ def get_latest_versions(self, name, stages=None):
...
-    def create_model_version(self, name, source, run_id):
+    def create_model_version(self, name, source, run_id, tags=None):
         """
         Create a new model version from given source or run ID.
 
         :param name: Name ID for containing registered model.
         :param source: Source path where the MLflow model is stored.
-        :param run_id: Run ID from MLflow tracking server that generated the model
+        :param run_id: Run ID from MLflow tracking server that generated the model.
+        :param tags: A list of :py:class:`mlflow.entities.model_registry.ModelVersionTag`
```
Comment (dbczumar) [inline mlflow/tracking/_model_registry/client.py:164]: For run tags, client-level APIs accept a dictionary of tags and converts them to a list of `RunTag` objects before passing them to the backend store (see https://github.com/mlflow/mlflow/blob/master/mlflow/tracking/_tracking_service/client.py#L88). I think we should do the same thing here.

File: mlflow/tracking/client.py [hunk 3]
```diff
@@ -353,15 +353,17 @@ def search_runs(self, experiment_ids, filter_string="", run_view_type=ViewType.A
     # Registered Model Methods
 
     @experimental
-    def create_registered_model(self, name):
+    def create_registered_model(self, name, tags=None):
         """
         Create a new registered model in backend store.
 
         :param name: Name of the new model. This is expected to be unique in the backend store.
+        :param tags: A list of :py:class:`mlflow.entities.model_registry.RegisteredModelTag`
```
Comment (dbczumar) [inline mlflow/tracking/client.py:361]: For run tags, client-level APIs accept a dictionary of tags and converts them to a list of `RunTag` objects before passing them to the backend store (see https://github.com/mlflow/mlflow/blob/master/mlflow/tracking/_tracking_service/client.py#L88). I think we should do the same thing here.

File: mlflow/tracking/client.py [hunk 4]
```diff
@@ -460,20 +462,45 @@ def get_latest_versions(self, name, stages=None):
...
     @experimental
-    def create_model_version(self, name, source, run_id):
+    def create_model_version(self, name, source, run_id, tags=None):
         """
         Create a new model version from given source or run ID.
 
         :param name: Name ID for containing registered model.
         :param source: Source path where the MLflow model is stored.
         :param run_id: Run ID from MLflow tracking server that generated the model
+        :param tags: A list of :py:class:`mlflow.entities.model_registry.ModelVersionTag`
```
Comment (dbczumar) [inline mlflow/tracking/client.py:498]: For run tags, client-level APIs accept a dictionary of tags and converts them to a list of `RunTag` objects before passing them to the backend store (see https://github.com/mlflow/mlflow/blob/master/mlflow/tracking/_tracking_service/client.py#L88). I think we should do the same thing here.

**Review by dbczumar — COMMENTED** [review]
@zhidongqu-db I tried this out and noticed that we throw exceptions for values with non-string types (e.g., ints) for model registry tag APIs, while `set_tag` and `create_run` APIs handle the casting of values to strings automatically (for example: https://github.com/mlflow/mlflow/blob/master/mlflow/tracking/_tracking_service/client.py#L202). Can we do this for MR tag APIs as well?

**Comment by zhidongqu-db** [comment]
> https://github.com/mlflow/mlflow/blob/master/mlflow/tracking/_tracking_service/client.py#L202

Good catch, I will do the cast for model registry tags as well.

**Comment by dbczumar** [comment]
@zhidongqu-db Can we add casting for the set of tags provided to `create_registered_model` and `create_model_version` as well? Currently, these methods fail with non-string values as follows:

```
client.create_registered_model(
    "mymod",
    tags={
        "foo": "bar",
        "bar": "foo7",
        "bar": 8888,
    })

Traceback (most recent call last):
  File "log.py", line 12, in <module>
    "bar": 8888,
  File "/Users/czumar/mlflow/mlflow/tracking/client.py", line 366, in create_registered_model
    return self._get_registry_client().create_registered_model(name, tags)
  File "/Users/czumar/mlflow/mlflow/tracking/_model_registry/client.py", line 46, in create_registered_model
    return self.store.create_registered_model(name, tags)
  File "/Users/czumar/mlflow/mlflow/store/model_registry/sqlalchemy_store.py", line 146, in create_registered_model
    _validate_registered_model_tag(tag.key, tag.value)
  File "/Users/czumar/mlflow/mlflow/utils/validation.py", line 121, in _validate_registered_model_tag
    _validate_length_limit("Registered model value", MAX_MODEL_REGISTRY_TAG_VALUE_LENGTH, value)
  File "/Users/czumar/mlflow/mlflow/utils/validation.py", line 155, in _validate_length_limit
    if len(value) > limit:
TypeError: object of type 'int' has no len()
```

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @zhidongqu-db !

**Pull request merged**